### DISCUSSION TITLE
Campaign object implementation

### DISCUSSION THREAD

**Discussion by karst** [OpenRCT2/OpenRCT2#20326] [Feature requests]
In this discussion I wanted present my ideas on how a "Campaign object" could be implemented.

* Campaigns will be their own file format. `.parkcampaign` based on the existing intransient `.parkobj` infrastructure.
* The `.parkcampaign` format would be a zip file allowing `.park`, `.sc6` and `.sc4`.
* There would be a `campaign.json` on the root of the zip file, this will have the following abilities:
  * Add a localised title to the campaign.
  * Add custom/localised discriminators between the scenarios. The way it's listed in the json will be the order they will be listed in the game.
  * Select a path to a scenario.
  * Modify/localise the scenario title.
  * Modify/localise the scenario description.
  * Add a campaign priority number. (Official campaigns will always have priority over custom).
  * Allow a custom banner (in the same location the Six Flags logo used to be), in indexed PNG format. This wil be displayed with all scenarios (and also the sandbox maps if these are implemented in their own tab).
  * ID: Same way as all other objects. Use the id naming scheme of `openrct2.campaign.whatever`/`<username>.campaign.<campaignname>`
  * Version: Same as objects.

The format could reject sandbox scenarios whenever those are implemented. Or you could add them as unlockables, in the sandbox selector grey'd out `"Complete <campaign name/scenario name> to unlock this sandbox map."`.

If you have more ideas let me know in the discussion.

**Comment by Gymnasiast** [comment]
I don’t think there is much reason to make them a separate format. IMO, they could easily be intransient objects, like audio (sound effects) objects are.

The idea of embedding saves with the file a good one. This could be made to work in a way similar to images, where objects can currently refer to vanilla DAT or G1 sprites or include their own.

**Reply by karst** [reply]
I mean my idea was that it'd essentially use most of the parkobj infrastructure, but for the user i think it makes more sense to have them displayed in a different folder (and the file extension with it for consitency and ease of use).

**Comment by ZehMatt** [comment]
I'm not sure I fully understand what you are asking here, are you talking about campaigns where the user has to complete scenarios in a certain order? To me also campaigns are something that keep track of your progress and also hand out rewards/abilities/unlocks that help you further progress the campaign, it would be also far more interesting if scenarios aren't locked out and you can start with any scenario you like but you would have to account for the unlocks maybe needed to complete the scenario. It has to be quite different from the normal scenario unlocking that would warrant this to be honest.

**Reply by karst** [reply]
I don't mind this idea, perhaps a flag for the scenarios what kind of unlock progression it has. For RCT1 scenarios a basic lockout would work, with RCT2 this isn't the case. Perhaps these can all be worked out over time, but leave open a field for the kind type of rewarding. Let's say certain rides only unlock in the campaign if a specific scenario is done. It would definitely make for very different types of scenarios than just "create an attraction park with this and this goal".

**Reply by ZehMatt** [reply]
That's what I was thinking, like unlocking rides is tied to the campaign profile rather than the played scenario and upon completion you would unlock new rides. Implementing all of this could be quite the challenge however.

**Reply by karst** [reply]
Perhaps it could be implemented by reserving a third list of research: researched, researchable and rewards, if a highscore is met within the campaign (where there game would tell the scenario list it'd unlock a new scenario) instead that scenario count would go up, and it'd unlock the first item in the list, etc.

**Comment by HtotheTML** [comment]
Is the idea that you'd be able to create these campaign organizers within the game itself?

**Reply by Gymnasiast** [reply]
I would expect that to be handled similarly to title sequences: with a plugin. External tools are also an option.

**Reply by HtotheTML** [reply]
So a plugin would be used to create the tabs on the left side of the scenario select window, but you could view them organized without it, or...?

**Comment by SpartanFrederic104** [comment]
What ideal image resolution would you propose for the campaign custom banners?

**Comment by kyphii** [comment]
(Apologies if I'm necro-ing, I can start a new discussion thread if desired)

I've been working on campaign object implementation here [#25336]. For a proof of concept, I've implemented it using a single JSON file to define the campaign. The scenarios aren't bundled in.

I can certainly change it to read a different file structure. Does the team have any preferences regarding the format? I see a few approaches:


**Approach 1 - Build off POC code, continue using JSON.**
PROs:
- Human readable, can easily be edited manually. Won't require any special tools or in-game campaign editor to be built for users to create their own campaigns.
- Flexible to additions. Data can be safely added to the file format without breaking previous versions of the game, so should need less code to handle different file versions if they are added.
- The code to read from JSON is already there, I've already done the work.

CONS:
- Insecure, gives a lot of power to the user to mess with the data in unexpected ways. The code needs plenty of sanity checks to ensure the game will handle things gracefully if the user creates strange or garbage campaign files.
- JSON can be picky regarding syntax, and doesn't fail particularly gracefully if the file has syntax errors. Users having trouble with their custom campaign files could become a frequently asked question.

**Approach 2 - Use JSON, but alias the file type, using .parkcampaign instead of.json.**
PROs:
- Same as above.

CONs:
- Same as above.

**Approach 3 - Use a custom file format utilizing DataSerialzer.**
PROs:
- More secure than JSON. A bespoke campaign editor, whether in-game or external, could limit user inputs to sane values, ensuring the created campaign works.

CONS:
- Will require more work, both to read/write new format, and more significantly create a bespoke in-game or external editor.
- Less flexible. If campaign format requires changes, extra code will be required to handle different versions.

**Approach 4 - Use a custom .ZIP file format, as detailed here.**
PROs:
- Has all advantages of JSON.
- Scenarios are bundled with the campaign, ensuring all files are present and the campaign appears as expected.

CONs:
- JSON disadvantages.
- Will require significantly more custom code to read.
- File size.
- How would this work for vanilla / base game campaigns? We would need to either start distributing the base game scenarios, or program an exception for official campaigns.
